### PR TITLE
Rewrite inline let to have lexical scope

### PR DESCRIPTION
**Pull request by mmun**
_EDIT: The syntax of {{let}} and {{#let}} has changed since this PR._

This PR reimplements inline let as an HTMLBars AST transform. There is no notion of inline let at runtime. It works by rewriting the let statements as nested blocks. For example,

``` hbs
<div>
  {{let a 1 b 2}}
  {{let a 11 b 12 c (add a b)}}
  {{a}} {{b}} {{c}}
  <div>
    {{let a 5}}
    {{a}} {{b}} {{c}}
  </div>
  {{let a (add a 10)}}
  {{a}} {{b}} {{c}}
</div>
```

is transformed to (excluding whitespace changes to make it easier to read):

``` hbs
<div>
  {{#let 1 2 as |a b|}}
    {{#let 11 12 (add a b) as |a b c|}}
      {{a}} {{b}} {{c}}
      <div>
        {{#let 5 as |a|}}
          {{a}} {{b}} {{c}}
        {{/let}}
      </div>
      {{#let (add a 10) as |a|}}
        {{a}} {{b}} {{c}}
      {{/let}}
    {{/let}}
  {{/let}}
</div>
```

Both templates generate the DOM

``` hbs
<div>
  11 12 3
  <div>
    5 12 3
  </div>
  21 12 3
</div>
```

### CODE REVIEW

**Comment by taras** [comment]
@mmun thank you! 

Ember 1.13 test is failing. In the current test suite, we're not testing inline let in 1.13. 

Should we exclude the AST transform from for Ember 1.13 or is it possible to make it work with 1.13?

**Comment by mmun** [comment]
Should be possible. :)

EDIT: Actually, it would be a pain to get it working in 1.13 so I disable the transform below 2.0.0.

**Comment by mmun** [comment]
I've changed the example in the PR message to demonstrate that bindings are created in parallel rather than series (E.g. the semantics are like LISP's `let` rather than `let*`).

**Comment by Robdel12** [comment]
This is pretty awesome 🎉

File: addon/initializers/register-let-helper.js [hunk 1]
```diff
@@ -17,22 +13,6 @@ export default {
         options.template.yield(params);
       }
```
Comment (cowboyd) [inline addon/initializers/register-let-helper.js:14]: For < 2.0 users, shouldn't there be an `else` with a warning that inline usage is not allowed?

Comment (mmun) [reply addon/initializers/register-let-helper.js:14]: We can do that.

**Comment by fivetanley** [comment]
is this glimmer 2 compatible?

**Comment by cowboyd** [comment]
An excellent question, is there a way that we can test it?

**Comment by Robdel12** [comment]
https://github.com/emberjs/ember.js/pull/14156 looks like we can build vs canary?

**Comment by cowboyd** [comment]
So looks like it's going to require some work to get it working on Glimmer. We can track it with #14, but I don't think it should hold this up getting merged.

**Comment by cowboyd** [comment]
I can't begin to express how happy this makes me. A truly lexical inline form is almost bringing tears to my eyes y'all. Not even joking. Thanks @mmun and @rwjblue making this happen.

**Comment by taras** [comment]
Absolutely, thanks guys so much for helping us make this work.

**Comment by mmun** [comment]
@cowboyd I have a follow up PR soon that adds the warning.

**Pull request merged**